aws/signer/v4/middleware.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// HandleFinalize compute the payload hash for the request payload
+func (m *ComputePayloadSHA256Middleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("unexpected request middleware type %T", in.Request)}
+	}
+
+	body, ok := req.Stream.(io.ReadSeeker)
+	if !ok {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("stream does not implement io.Seeker, got %T", req.Stream)}
+	}
+
+	_, err = body.Seek(0, io.SeekStart)
+	if err != nil {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("failed to seek body to start, %w", err)}
+	}
+
+	hash := sha256.New()
+	_, err = io.Copy(hash, body)
+	if err != nil {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("failed to compute payload hash, %w", err)}
+	}
+
+	_, err = body.Seek(0, io.SeekStart)
+	if err != nil {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("failed to seek body to start, %w", err)}
+	}
+
+	ctx = SetPayloadHash(ctx, hex.EncodeToString(hash.Sum(nil)))
+
+	return next.HandleFinalize(ctx, in)
+}
+
+// SignHTTPRequestMiddleware is a `FinalizeMiddleware` implementation for SigV4 HTTP Signing
+type SignHTTPRequestMiddleware struct {
+	signer HTTPSigner
+}
+
+// NewSignHTTPRequestMiddleware constructs a SignHTTPRequestMiddleware using the given Signer for signing requests
+func NewSignHTTPRequestMiddleware(signer HTTPSigner) *SignHTTPRequestMiddleware {
+	return &SignHTTPRequestMiddleware{signer: signer}
+}
+
+// ID is the SignHTTPRequestMiddleware identifier
+func (s *SignHTTPRequestMiddleware) ID() string {
+	return "SigV4SignHTTPRequestMiddleware"
+}
+
+// HandleFinalize will take the provided input and sign the request using the SigV4 authentication scheme
+func (s *SignHTTPRequestMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return middleware.FinalizeOutput{}, metadata, &SigningError{Err: fmt.Errorf("unexpected request middleware type %T", in.Request)}
+	}
+
+	signingMetadata := GetSigningMetadata(ctx)
+	payloadHash := GetPayloadHash(ctx)
+	if len(payloadHash) == 0 {
+		return middleware.FinalizeOutput{}, metadata, &SigningError{Err: fmt.Errorf("computed payload hash missing from context")}
+	}
+
+	err = s.signer.SignHTTP(ctx, req.Request, payloadHash, signingMetadata.SigningName, signingMetadata.SigningRegion, sdk.NowTime())
+	if err != nil {
+		return middleware.FinalizeOutput{}, metadata, &SigningError{Err: fmt.Errorf("failed to sign http request, %w", err)}
```

Comment:
could just use `out` for the return instead of middleware.FinalizeOutput{} similar to metadata